cmd/earthly/settings_test.go

```diff
@@ -0,0 +1,354 @@
+				localRegistry: "",
+			},
+		},
+		{
+			"Debugger host and port specified",
+			config.GlobalConfig{
+				BuildkitHost:      "",
+				DebuggerHost:      "tcp://127.0.0.1:1234",
+				DebuggerPort:      5678,
+				LocalRegistryHost: "",
+			},
+			noopArgs,
+			addresses{
+				buildkit:      buildkitd.DockerAddress,
+				debugger:      "tcp://127.0.0.1:1234",
+				localRegistry: "",
+			},
+		},
+		{
+			"No Config, buildkit and debugger on CLI",
+			config.GlobalConfig{
+				BuildkitHost:      "tcp://not-this-bk:1234",
+				DebuggerHost:      "tcp://not-this-db:1234",
+				DebuggerPort:      config.DefaultDebuggerPort,
+				LocalRegistryHost: "",
+			},
+			[]string{"", "--buildkit-host", "tcp://ok-bk:42", "--debugger-host", "tcp://ok-db:43"},
+			addresses{
+				buildkit:      "tcp://ok-bk:42",
+				debugger:      "tcp://ok-db:43",
+				localRegistry: "",
+			},
+		},
+		{
+			"Remote Local in config, no CLI, validate registry host",
+			config.GlobalConfig{
+				BuildkitHost:      "tcp://127.0.0.1:8372",
+				DebuggerHost:      "",
+				DebuggerPort:      config.DefaultDebuggerPort,
+				LocalRegistryHost: "tcp://127.0.0.1:8371",
+			},
+			noopArgs,
+			addresses{
+				buildkit:      "tcp://127.0.0.1:8372",
+				debugger:      fmt.Sprintf("tcp://127.0.0.1:%v", config.DefaultDebuggerPort),
+				localRegistry: "tcp://127.0.0.1:8371",
+			},
+		},
+		{
+			"Remote remote in config, no CLI, skip validate registry host",
+			config.GlobalConfig{
+				BuildkitHost:      "tcp://my-cool-host:8372",
+				DebuggerHost:      "",
+				DebuggerPort:      config.DefaultDebuggerPort,
+				LocalRegistryHost: "this-is-not-a-url",
+			},
+			noopArgs,
+			addresses{
+				buildkit:      "tcp://my-cool-host:8372",
+				debugger:      fmt.Sprintf("tcp://my-cool-host:%v", config.DefaultDebuggerPort),
+				localRegistry: "this-is-not-a-url",
+			},
+		},
+		{
+			"Local in config, no CLI, validate registry host",
+			config.GlobalConfig{
+				BuildkitHost:      "docker-container://my-cool-container",
+				DebuggerHost:      "",
+				DebuggerPort:      config.DefaultDebuggerPort,
+				LocalRegistryHost: "tcp://127.0.0.1:8371",
+			},
+			noopArgs,
+			addresses{
+				buildkit:      "docker-container://my-cool-container",
+				debugger:      fmt.Sprintf("tcp://127.0.0.1:%v", config.DefaultDebuggerPort),
+				localRegistry: "tcp://127.0.0.1:8371",
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		ctx := context.Background()
+
+		var logs strings.Builder
+		var trash strings.Builder
+
+		logger := conslogging.Current(conslogging.NoColor, conslogging.DefaultPadding, false)
+		logger = logger.WithWriter(&logs)
+
+		earthlyApp := newEarthlyApp(ctx, logger)
+		earthlyApp.cfg = &config.Config{Global: tt.config}
+		earthlyApp.cliApp.Writer = &trash    // Just chuck the help output
+		earthlyApp.cliApp.ErrWriter = &trash // All of it, we dont care
+
+		// Before is called at about the time that we would parse these, plus it a nice place to hook.
```

Comment:
interesting hook. This test almost feels like an integration test that could be done via earthly-in-earthly which might provide more test coverage; but that would definitely take more execution time. I think it's fine here though.

Reply:
Yeah. I also would like too be able to debug these.

Reply:
Little more clarification:

I only hooked here because it was easier than building the `cli.Context` and all the stuff along with it (like a `FlagSet`). Waaay easier to allow the CLI framework to give me one.